File: day5/day5.py
from dataclasses import dataclass


def parse_input(filename) -> list[list[int]]:
    ints = []
    with open(filename, 'r+') as fi:
      for line in fi.readlines():
        if ":" in line:
          ints.append([])
        found = list(map(int,
          filter(lambda x: len(list(
            filter(str.isdigit, x)
          )) > 0,
          line.split(" "))
        ))
        if len(found) > 0:
          ints[-1].extend(found)
    return ints

@dataclass
class TRange:
  left: int
  right: int
  index: int

  def __in__(self, other: int):
    return self.left <= other <= self.right

  def intersect(self, a):
    return TRange(left=max(self.left, a.left), right=min(self.right, a.right), index=a.index)

  def __str__(self):
    return f"TRange({self.left}, {self.right})"

@dataclass
class TUnit:
  destination: TRange
  source: TRange

  def translate(self, tr):
    out = TRange(tr.left, tr.right, tr.index)
    out.left += self.destination.left - self.source.left
    out.right += self.destination.right - self.source.right
    return out 


def solve_a(nums: list[list[int]]):
  queue = [*map(lambda x: TRange(left=x, right=x, index=0), nums[0])]
  groups = [[TUnit(
    destination=TRange(left=nums[i][j], right=nums[i][j] + nums[i][j+2], index=i),
    source=TRange(left=nums[i][j+1], right=nums[i][j+1] + nums[i][j+2], index=i),
  ) for j in range(0, len(nums[i]), 3)] for i in range(1, len(nums))]
  while queue[0].index < len(groups):
    front = queue[0]
    queue = queue[1:]
    group = groups[front.index]
    front.index += 1
    matched = False
    for tr in group:
      inter = front.intersect(tr.source)
      if inter.left <= inter.right:
        queue.append(tr.translate(inter))
        matched = True
    if not matched:
      queue.append(front)
  return min(queue, key=lambda x: x.left).left


def prepare_groups(nums):
  groups = [
    [TUnit(
      destination=TRange(left=nums[i][j], right=nums[i][j]+nums[i][j+2]-1, index=i),
      source=TRange(left=nums[i][j+1], right=nums[i][j+1]+nums[i][j+2]-1, index=i),
      ) for j in range(0, len(nums[i]), 3)
    ] for i in range(1, len(nums))
  ]
  for i in range(len(groups)):
    groups[i] = sorted(groups[i], key=lambda x: x.source.right)
    groups[i] = sorted(groups[i], key=lambda x: x.source.left)
  return groups


def solve_b_next_queue(queue: list[TRange], group: list[TUnit]):
  for _ in range(len(queue)):
    front = queue[0]
    queue = queue[1:]
  
    front.index += 1
    left_min = front.left
    # make sure the input range is covered
    for tr in filter(lambda x: x.source.right >= front.left, group):
      inter = front.intersect(tr.source)
      if inter.left <= inter.right:
        queue.append(tr.translate(inter))
        if left_min < inter.left:
          queue.append(TRange(left=left_min, right=inter.left-1, index=inter.index))
        left_min = max(left_min, inter.right) + 1
    if left_min < front.right:
      queue.append(TRange(left=left_min, right=front.right, index=front.index))
  # merge overlap in resulting queue
  queue.sort(key=lambda x: x.right)
  queue.sort(key=lambda x: x.left)
  for i in range(len(queue)-1,0,-1):
    inter = queue[i].intersect(queue[i-1])
    if inter.left <= inter.right or queue[i].right + 1 == queue[i-1].left or queue[i].left == queue[i-1].right+1:
      queue[i].left = min(queue[i].left, queue[i-1].left)
      queue[i].right = max(queue[i].right, queue[i-1].right)
      queue = queue[:i-1] + queue[i:]

  return queue


def solve_b(nums: list[list[int]]) -> int:
  queue = [
    TRange(left=nums[0][i], right=nums[0][i]+nums[0][i+1]-1, index=0)
    for i in range(0, len(nums[0]), 2)
  ]
  groups = prepare_groups(nums)
  for group in groups:
    queue = solve_b_next_queue(queue, group)
  return min(queue, key=lambda x: x.left).left


def full_test():
  aoc_input = parse_input("test.txt")
  ans = solve_a(aoc_input)
  assert ans == 35, f"Expected 35, got {ans}"
  ans = solve_b(aoc_input)
  assert ans == 46, f"Expected 46, got {ans}"

  aoc_input = parse_input("input.txt")
  ans = solve_a(aoc_input)
  assert ans == 240320250, f"Expected 240320250, got {ans}"
  ans = solve_b(aoc_input)
  assert ans == 28580589, f"Expected 11787590, got {ans}"

if __name__ == "__main__":
  full_test()

File: day5/day5.cpp
#include "day5.h"
#include "../aoc/aoc.h"

#include <vector>
#include <queue>

namespace day5 {
template<typename T>
void print_vec(const std::vector<T>& v) {
	std::cout << "[";
	for (int a : v) {
		std::cout << a << ", ";
	}
	std::cout << "]" << std::endl;
}

std::vector<std::vector<int64_t>> to_numbers(const std::string& s, const std::string& split_on) {
	std::vector<std::vector<int64_t>> v{std::vector<int64_t>{}};
	for (auto it = s.begin(); it != s.end();) {
		if (split_on.find(*it) != std::string::npos) {
			v.push_back(std::vector<int64_t>{});
		}
		if (::isdigit(*it)) {
			auto scan = it;
			while (::isdigit(*(scan++)));
			v.back().push_back(stoll(std::string{it, scan}));
			it = scan;
		} else {
			it++;
		}
	}
	return v;
}

struct TranslationUnit {
	size_t index;
	int64_t value;
};

int64_t solve_a(const std::string& input) {
	auto groups = to_numbers(input, ":");
	groups.erase(groups.begin()); // first vector is empty from "seeds: ..."
	auto seeds = groups[0];
	groups.erase(groups.begin()); // seeds

	std::queue<TranslationUnit> search;
	for (auto seed : seeds) {
		search.push({0, seed});
	}

	while (!search.empty() && search.front().index < groups.size()) {
		auto tu = search.front();
		search.pop();
		
		bool matched = false;
		auto tg = groups[tu.index];
		tu.index++;
		for (size_t i = 0; i < tg.size(); i+=3) { // groups of 3: destination source range
			auto destination = tg[i], source = tg[i+1], range = tg[i+2]; 
			auto diff = tu.value - source;
			if (diff > 0 && diff < range) {
				matched = true;
				search.push({tu.index, destination + diff});
			}
		}
		if (!matched) search.push(tu);
	}

	int64_t min = 1LL << 63 ^ -1;
	while (!search.empty()) {
		min = std::min(min, search.front().value);
		search.pop();
	}
	return min;
}

struct TRange {
	size_t index;
	ssize_t left;
	ssize_t right;
};

bool has_intersection(std::pair<uint64_t, uint64_t> a, std::pair<uint64_t, uint64_t> b) {
	return b.first <= a.first && a.first + a.second <= b.first + b.second
	|| b.first >= a.first && b.first <= a.first + a.second
	|| b.first + b.second >= a.first && b.first + b.second <= a.first + a.second;
}

std::vector<TRange> split(std::vector<int64_t>& group, TRange tr) {
	std::vector<TRange> matched;
	std::queue<TRange> unprocessed;

	tr.index++; // for the next intersection
	for (int i = 0; i < group.size(); i += 3) {
		auto destination = group[0], source = group[1], range = group[2];
		// TODO: translating TRange to left right
		auto source_end = source + range, dest_end = destination + range;
		
	}
	
	// verify there entire range is covered in match, sort
	// maintains relative [first,second] ordering. e.g. [1,3] < [1,4]
	std::stable_sort(matched.begin(), matched.end(),
		[](const TRange a, const TRange b) {
			return a.right < b.right;
		 });
	std::stable_sort(matched.begin(), matched.end(),
		[](const TRange a, const TRange b) { return a.left < b.left; });
	auto curr = tr;
	for (size_t j = 0; j < matched.size(); j++) {
		// find any case where -----....--------....------
		//                          ^curr
		// curr refers to an uncovered region
		auto match = matched[j];
		if (curr.left < match.left) {
			// mismatch found
			TRange mismatch = {tr.index, curr.left, match.left - curr.left};
			matched.push_back(mismatch);
			curr.left = mismatch.left + mismatch.right;
		} else {
			curr.left = match.left;
		}
		curr.right = tr.right - curr.left;
	}
	return matched;
}

int64_t solve_b(const std::string& input) {
	std::cout << "SOLVE B" << std::endl;
	auto groups = to_numbers(input, ":");
	groups.erase(groups.begin()); // first vector is empty from "seeds: ..."
	auto seeds = groups[0];
	groups.erase(groups.begin()); // seeds

	std::queue<TRange> search;
	for (size_t i = 0; i < seeds.size(); i+=2) {
		search.push({0, seeds[i], seeds[i+1]});
	}

	// assume this represents a DAG; a solution can be found w.o backtracking
	while (!search.empty() && search.front().index < groups.size()) {
		auto front = search.front();
		search.pop();
		auto tg = groups[front.index];
		front.index++;
		for (auto match : split(tg, front)) {
			search.push(match);
		}
	}

	int64_t min = 1LL << 63 ^ -1;
	while (!search.empty()) {
		min = std::min(min, search.front().left);
		search.pop();
	}
	return min;
}
}
